language/diem-vm/src/diem_transaction_validator.rs

```diff
@@ -2,58 +2,159 @@
-            "Adapter created for Validation"
-        );
-        Self(DiemVMImpl::new(state))
+        Ok(())
     }
 
-    pub fn init_with_config(
-        version: DiemVersion,
-        on_chain_config: VMConfig,
-        publishing_option: VMPublishingOption,
-    ) -> Self {
-        info!("Adapter restarted for Validation");
-        DiemVMValidator(DiemVMImpl::init_with_config(
-            version,
-            on_chain_config,
-            publishing_option,
-        ))
+    fn get_gas_price<S: MoveResolver>(&self, txn: &SignedTransaction, remote_cache: &S) -> Result<u64, VMStatus> {
+        let gas_price = txn.gas_unit_price();
+        let currency_code = get_gas_currency_code(txn)?;
+
+        let normalized_gas_price = match get_currency_info(&currency_code, remote_cache) {
+            Ok(info) => info.convert_to_xdx(gas_price),
+            Err(err) => {
+                return Err(err);
+            }
+        };
+
+        Ok(normalized_gas_price)
+    }
+
+    fn run_prologue<S: MoveResolver>(
+        &self,
+        session: &mut Session<S>,
+        transaction: &SignatureCheckedTransaction,
+        log_context: &AdapterLogSchema,
+    ) -> Result<(), VMStatus> {
+        let currency_code = get_gas_currency_code(transaction)?;
+        let txn_data = TransactionMetadata::new(transaction);
+        match transaction.payload() {
+            TransactionPayload::Script(_) => {
+                self.0.check_gas(&txn_data, log_context)?;
+                self.0
+                    .run_script_prologue(session, &txn_data, &currency_code, log_context)
+            }
+            TransactionPayload::ScriptFunction(_) => {
+                // gate the behavior until the Diem version is ready
+                if self.0.get_diem_version()? < DIEM_VERSION_2 {
+                    return Err(VMStatus::Error(StatusCode::FEATURE_UNDER_GATING));
+                }
+                // NOTE: Script and ScriptFunction shares the same prologue
+                self.0.check_gas(&txn_data, log_context)?;
+                self.0
+                    .run_script_prologue(session, &txn_data, &currency_code, log_context)
+            }
+            TransactionPayload::Module(_module) => {
+                self.0.check_gas(&txn_data, log_context)?;
+                self.0
+                    .run_module_prologue(session, &txn_data, &currency_code, log_context)
+            }
+            TransactionPayload::WriteSet(_cs) => {
+                self.0
+                    .run_writeset_prologue(session, &txn_data, log_context)
+            }
+        }
+    }
+
+    fn should_restart_execution(vm_output: &TransactionOutput) -> bool {
+        let new_epoch_event_key = diem_types::on_chain_config::new_epoch_event_key();
+        vm_output
+            .events()
+            .iter()
+            .any(|event| *event.key() == new_epoch_event_key)
+    }
+
+    fn execute_single_transaction<S: MoveResolver + StateView>(
```

Comment:
With the current model, the client of the VMAdapter trait still has to implement the bulk of the transaction logic by itself. By this I mean `DiemVM::execute_user_transaction` and a bunch of functions it calls into, including the failed transaction clean-up logic. IMO that part of the code is the most difficult to get right and maintain, and I would love if we can come up with a robust generic implementation for it.